### ISSUE TITLE
`fetchAuthSession` throws SignedOutException prior to refresh token expiration.

### ISSUE THREAD

**Issue by sourabhguptazeil** [aws-amplify/amplify-flutter#2808]
### Login methods are affected

1. Login with email
2. Sign in with google
3. Sign in with Apple

### The expiration time set in Cognito for all tokens (access, id, refresh)

1. Refresh token expiry is 180 days
2. Access token expiry is 1 day

### How long after login the exception is occurring

Generally it occurs when 24 hours are passed

### How do we access the token

```
Future<String?> fetchIdentityToken() async {
    final result = await Amplify.Auth.fetchAuthSession(
          options: const FetchAuthSessionOptions());
   String? identityId = (result as CognitoAuthSession)
          .userPoolTokensResult
          .valueOrNull
          ?.idToken
          .raw;
   return identityId;
}
```

### What exception are they seeing
```
  "isSignedIn": false,
  "userSub": "AWSResult [null, SignedOutException {\n  \"message\": \"No user is currently signed in\"\n}, AWSResultType.error]",
  "userPoolTokens": "AWSResult [null, SignedOutException {\n  \"message\": \"No user is currently signed in\"\n}, AWSResultType.error]",
  "credentials": "AWSResult [null, SessionExpiredException {\n  \"message\": \"The AWS credentials could not be retrieved\",\n  \"recoverySuggestion\": \"Invoke Amplify.Auth.signIn to re-authenticate the user\",\n  \"underlyingException\": \"NotAuthorizedException {\\n  message=Unauthenticated access is not supported for this identity pool.,\\n}\"\n}, AWSResultType.error]",
  "identityId": "AWSResult [null, SessionExpiredException {\n  \"message\": \"The AWS credentials could not be retrieved\",\n  \"recoverySuggestion\": \"Invoke Amplify.Auth.signIn to re-authenticate the user\",\n  \"underlyingException\": \"NotAuthorizedException {\\n  message=Unauthenticated access is not supported for this identity pool.,\\n}\"\n}, AWSResultType.error]"
}
```

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [X] Auth
- [ ] Authenticator
- [ ] DataStore
- [ ] Storage

### Platforms

- [X] iOS
- [X] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.7.9

### Amplify Flutter Version

1.0.0-next.7

### Deployment Method

Custom Pipeline

**Comment by Jordan-Nelson** [comment]
Hi @sourabhguptazeil - Thanks for opening the issue.

> Generally it occurs when 24 hours are passed

Can you elaborate on this? How frequently does it occur? Are you able to reproduce this fairly consistently?

Are you seeing this on physical devices, emulators, or both?

**Comment by sourabhguptazeil** [comment]
@Jordan-Nelson 
1. Issue occurs all the time.
2. Issue is reproducible consistently.
3. Issue is reproducible on both emulator and physical device.

**Comment by Jordan-Nelson** [comment]
@sourabhguptazeil - I am unable to reproduce this. I have created a Cognito back end with the same token expiration settings that you have described. I signed into the app (on iOS) over 36 hours ago and my session has not been interrupted.

What the Cognito back end created with the Amplify CLI?

Can you share a screenshot of your App Client setting from the Cognito console so that I can make sure we have the exact same configuration? You can omit the app client ID, but please make sure it is the same app client ID that is on your Amplify Config. See below for an example.

![Screen Shot 2023-04-05 at 10 26 21 AM](https://user-images.githubusercontent.com/20613561/230111960-b0bd357e-2854-4309-8e79-8c54b1d66796.png)

**Comment by sourabhguptazeil** [comment]
@Jordan-Nelson 
## What the Cognito back end created with the Amplify CLI?

Cognito backend was created manually without amplify CLI because we only wanted to integrate authentication sdk on front end side.

I followed `manual configuration` from this section https://docs.amplify.aws/lib/auth/getting-started/q/platform/flutter/#set-up-backend-resources
<img width="545" alt="Screenshot 2023-04-06 at 3 34 28 AM" src="https://user-images.githubusercontent.com/108905944/230131096-630ca470-2ff4-47a9-9892-05bb3b51c6f4.png">

```
{
    "UserAgent": "aws-amplify-cli/2.0",
    "Version": "1.0",
    "auth": {
        "plugins": {
            "awsCognitoAuthPlugin": {
                "IdentityManager": {
                    "Default": {}
                },
                "CredentialsProvider": {
                  "CognitoIdentity": {
                      "Default": {
                          "PoolId": "<This is identity pool id created using cognito pool and client id>",
                          "Region": "ap-southeast-2"
                      }
                  }
                },
                "CognitoUserPool": {
                    "Default": {
                        "PoolId": "<Pool id goes here>",
                        "AppClientId": "<Client id goes here>",
                        "AppClientSecret": "<Secrete Goes here>",
                        "Region": "ap-southeast-2"
                    }
                },
                "Auth": {
                    "Default": {
                        "OAuth": {
                            "WebDomain": "xyz.auth.ap-southeast-2.amazoncognito.com",
                            "AppClientId": "<Client id goes here>",
                            "AppClientSecret": "<Secrete Goes here>",
                            "SignInRedirectURI": "<SignIn Redirect>",
                            "SignOutRedirectURI": "SignOut Redirect",
                            "Scopes": ["email","openid","profile","aws.cognito.signin.user.admin"]
                        },
                        "authenticationFlowType": "USER_SRP_AUTH",
                        "socialProviders": [
                            "GOOGLE", "APPLE"
                        ]
                    }
                }
            }
        }
    }
  }
```
### Difference
1. I have extra attribute `AppClientSecret`. If I don't add the secret then following exception occurs
```
I/flutter (11264): ERROR | SignInStateMachine | Emitted error: NotAuthorizedException {
I/flutter (11264):   "message": "Client <CLIENT_ID> is configured with secret but SECRET_HASH was not received",
I/flutter (11264):   "underlyingException": "NotAuthorizedException {\n  message=Client <CLIENT_ID> is configured with secret but SECRET_HASH was not received,\n}"
```

## Cognito console configuration from AWS console
<img width="425" alt="Screenshot 2023-04-06 at 3 16 56 AM" src="https://user-images.githubusercontent.com/108905944/230126617-4a3b2b4d-ab49-477e-919b-b27b8b904ce0.png">

## Identity pool configuration from AWS console
<img width="583" alt="Screenshot 2023-04-06 at 3 29 07 AM" src="https://user-images.githubusercontent.com/108905944/230129955-5b06cc2a-9c4d-43d1-867a-1bad34e9de5a.png">

**Comment by dnys1** [comment]
Hi @sourabhguptazeil - this does appear to be an issue with our handling of client secret. Nothing you've done wrong - I've put up a [PR](https://github.com/aws-amplify/amplify-flutter/pull/2819) which fixes this issue and adds some e2e test coverage.

**Comment by sourabhguptazeil** [comment]
@dnys1 Can front end code work without specifying the client secret with Cognito dev pool where secret is configured.

Ideally we would like to remove client secret from front end side.

**Comment by dnys1** [comment]
Indeed, frontend code can work without a client secret and this is generally the preferred strategy for frontend clients. Typically, a client secret is reserved for backend servers.

I believe removing the client secret should also resolve this issue if that is preferred.

**Comment by sourabhguptazeil** [comment]
@dnys1  Just to confirm

The [PR](https://github.com/aws-amplify/amplify-flutter/pull/2819) will fix the following exception once client secret is removed from the front end

```
I/flutter (11264): ERROR | SignInStateMachine | Emitted error: NotAuthorizedException {
I/flutter (11264):   "message": "Client <CLIENT_ID> is configured with secret but SECRET_HASH was not received",
I/flutter (11264):   "underlyingException": "NotAuthorizedException {\n  message=Client <CLIENT_ID> is configured with secret but SECRET_HASH was not received,\n}"

```

**Comment by dnys1** [comment]
This PR will fix the issue when the client secret is present. If the client secret is removed, there should not currently be any issue.

**Comment by sourabhguptazeil** [comment]
@dnys1 Actually following exception occurs when I remove client secret from front end code but Cognito dev pool is configured with secret

```
I/flutter (11264): ERROR | SignInStateMachine | Emitted error: NotAuthorizedException {
I/flutter (11264):   "message": "Client <CLIENT_ID> is configured with secret but SECRET_HASH was not received",
I/flutter (11264):   "underlyingException": "NotAuthorizedException {\n  message=Client <CLIENT_ID> is configured with secret but SECRET_HASH was not received,\n}"

```

**Comment by dnys1** [comment]
I apologize for the confusion. By "removing the client secret", I had assumed you meant removing it from your backend and frontend code. If Cognito expects a client secret, then this is exception is correct. Your options are to:
1. Disable the client secret in the app client in Cognito and remove it from the frontend
2. Create a new app client without a client secret and remove it from the frontend
3. Leave the frontend untouched and wait for the fix to be released

**Comment by dnys1** [comment]
The fix for this has been merged and will be released soon. Please let us know if you experience any more issues and thank you for your patience resolving this.

**Comment by zeiladmin** [comment]
@dnys1 Thanks very much for looking into the issue and fix it so quickly.

@Jordan-Nelson Thanks very much for triaging the issue and for your guidance.

**Comment by sourabhguptazeil** [comment]
@dnys1 Thanks very much for looking into the issue and fix it so quickly.

@Jordan-Nelson Thanks very much for triaging the issue and for your guidance.

**Comment by Jordan-Nelson** [comment]
The fix has been included in the first stable release of v1.0. No change should be required.

**Comment by holographix** [comment]
Sorry to revamp this issue, but is it possible that I'm facing the same problem running the v. 2.4.0? I've seen this going on an issue on the js lib repo: https://github.com/aws-amplify/amplify-js/issues/1218
Anybody has  had this issue as well?

`      I/flutter ( 6795): [AuthService] NotAuthorizedServiceException {
I/flutter ( 6795):   "message": "Client XXXXXXYYYYZZZZ is configured with secret but SECRET_HASH was not received",
I/flutter ( 6795):   "underlyingException": "NotAuthorizedException {\n  message=Client XXXXXXYYYYYZZZZ is configured with secret but SECRET_HASH was not received,\n}"
I/flutter ( 6795): }

`
